[JIT] Enable @unused syntax for ignoring properties

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45234 [JIT] Add recursive scripting for class type module attributes
* #45262 [JIT] Add API for ignoring arbitrary module attributes
* **#45261 [JIT] Enable @unused syntax for ignoring properties**
* #45233 [JIT] Add test for ignored class type property
* #45232 [JIT] Update docs for recently added features

**Summary**
This commit enables `@unused` syntax for ignoring
properties. Inoring properties is more intuitive with this feature enabled.
`@ignore` is not supported because class type properties cannot be
executed in Python (because they exist only as TorchScript types) like
an `@ignored` function and module properties that cannot be scripted
are not added to the `ScriptModule` wrapper so that they
may execute in Python.

**Test Plan**
This commit updates the existing unit tests for class type and module
properties to test properties ignored using `@unused`.

Differential Revision: [D23971881](https://our.internmc.facebook.com/intern/diff/D23971881)

Commit:
[JIT] Enable @ignore and @unused syntax for ignoring properties

**Summary**
This commit enables `@ignore` and `@unused` syntax for ignoring
properties, but works only if these decorators are applied to a function
before the @property decorator. Despite this, ignoring properties is
more intuitive with this feature enabled.

**Test Plan**
This commit updates the existing unit tests for class type and module
properties to test properties ignored using `@ignore` and `@unused`.
Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "[JIT] Enable @ignore and @unused syntax for ignoring properties"


**Summary**
This commit enables `@ignore` and `@unused` syntax for ignoring
properties, but works only if these decorators are applied to a function
before the @property decorator. Despite this, ignoring properties is
more intuitive with this feature enabled.

**Test Plan**
This commit updates the existing unit tests for class type and module
properties to test properties ignored using `@ignore` and `@unused`.
Summary:

[ghstack-poisoned]

Commit:
Update on "[JIT] Enable @ignore and @unused syntax for ignoring properties"


**Summary**
This commit enables `@ignore` and `@unused` syntax for ignoring
properties, but works only if these decorators are applied to a function
before the @property decorator. Despite this, ignoring properties is
more intuitive with this feature enabled.

**Test Plan**
This commit updates the existing unit tests for class type and module
properties to test properties ignored using `@ignore` and `@unused`.
Summary:

[ghstack-poisoned]

eellison commented on test/jit/test_class_type.py:
```diff
@@ -1180,6 +1180,15 @@ def attr(self) -> int:
             def unsupported(self) -> int:
                 return sum([self.a])
 
+            @property
```
Do you know would it take to support the unused decorator. and then property ? That seems more intuitive to me. With the current API I would expect the property to exist but the getter be an ignored function, whereas we're just completely dropping it

```
            @torch.jit.unused
            @property
            def unsupported_2(self) -> int:
                return sum([self.a])

            @unsupported_2.setter
            def unsupported_2(self, value):
                self.a = sum([self.a])
```

SplitInfinity replied:
So this is the thing that I tried to do in my diff for properties on modules. But there are backwards compatibility issues because using `@unused` or `@ignored` like this produces an error in older versions of PyTorch.

The concrete case we ran into is that some internal library is using pytorch 1.4 in their OSS tree and pytorch from master in fbcode.

eellison replied:
Trying to make new features & syntax work across older versions of pytorch is going to limit the clarity and intuitiveness of the APIs we make. it's also not a very common use case. It's good to offer escape hatches here (like __jit_ignored_attributes__) but I don't think we should overly taylor APIs to it

SplitInfinity replied:
Oh, I thought you were meant that we should introduce a decorator-based syntax and get rid of the `__ignored_attributes__` escape hatch. But if we're going to have both, I agree that it's better to allow people to write:
```
@torch.jit.unused
@property 
...
```

I'll update the PR.

Commit:
Update on "[JIT] Enable @ignore and @unused syntax for ignoring properties"


**Summary**
This commit enables `@ignore` and `@unused` syntax for ignoring
properties, but works only if these decorators are applied to a function
before the @property decorator. Despite this, ignoring properties is
more intuitive with this feature enabled.

**Test Plan**
This commit updates the existing unit tests for class type and module
properties to test properties ignored using `@ignore` and `@unused`.
Summary:

[ghstack-poisoned]

Commit:
Update on "[JIT] Enable @unused syntax for ignoring properties"


**Summary**
This commit enables `@unused` syntax for ignoring
properties. Inoring properties is more intuitive with this feature enabled.
`@ignore` is not supported because class type properties cannot be
executed in Python (because they exist only as TorchScript types) like
an `@ignored` function and module properties that cannot be scripted
are not added to the `ScriptModule` wrapper so that they
may execute in Python.

**Test Plan**
This commit updates the existing unit tests for class type and module
properties to test properties ignored using `@unused`.

[ghstack-poisoned]

Commit:
Update on "[JIT] Enable @unused syntax for ignoring properties"


**Summary**
This commit enables `@unused` syntax for ignoring
properties. Inoring properties is more intuitive with this feature enabled.
`@ignore` is not supported because class type properties cannot be
executed in Python (because they exist only as TorchScript types) like
an `@ignored` function and module properties that cannot be scripted
are not added to the `ScriptModule` wrapper so that they
may execute in Python.

**Test Plan**
This commit updates the existing unit tests for class type and module
properties to test properties ignored using `@unused`.

[ghstack-poisoned]

gmagogsfm reviewed (approved):
Change looks good, just curious is there any docs mentioning that @unused rather than @ignored should be used on properties? If user mistakenly uses @ignored on a property, what's going to happen? Will they get a good error message?

Commit:
Update on "[JIT] Enable @unused syntax for ignoring properties"


**Summary**
This commit enables `@unused` syntax for ignoring
properties. Inoring properties is more intuitive with this feature enabled.
`@ignore` is not supported because class type properties cannot be
executed in Python (because they exist only as TorchScript types) like
an `@ignored` function and module properties that cannot be scripted
are not added to the `ScriptModule` wrapper so that they
may execute in Python.

**Test Plan**
This commit updates the existing unit tests for class type and module
properties to test properties ignored using `@unused`.

[ghstack-poisoned]

Commit:
Update on "[JIT] Enable @unused syntax for ignoring properties"


**Summary**
This commit enables `@unused` syntax for ignoring
properties. Inoring properties is more intuitive with this feature enabled.
`@ignore` is not supported because class type properties cannot be
executed in Python (because they exist only as TorchScript types) like
an `@ignored` function and module properties that cannot be scripted
are not added to the `ScriptModule` wrapper so that they
may execute in Python.

**Test Plan**
This commit updates the existing unit tests for class type and module
properties to test properties ignored using `@unused`.

[ghstack-poisoned]

eellison commented on test/jit/test_class_type.py:
```diff
@@ -1180,6 +1180,15 @@ def attr(self) -> int:
             def unsupported(self) -> int:
                 return sum([self.a])
 
+            @torch.jit.unused
```
The convention of `torch.jit.unused` is that the function can occur as a call within a script'd. function but will error if used.  Does. that work here ?

SplitInfinity replied:
If a function tries to call this, an error will be produced that says that there is no such attribute. Not quite as helpful as the usual `unused` error, but the behaviour is closed to `unused` than to `ignore`.

eellison replied:
Okay, sgtm.

eellison commented on test/jit/test_class_type.py:
```diff
@@ -1167,7 +1167,7 @@ def free_function(x: int) -> int:
         @torch.jit.script
         class Properties(object):
-            __ignored_properties__ = ["unsupported"]
+            __jit_unused_properties__ = ["unsupported"]
```
Ideally we don't need to add another class attribute API and `__jit_ignored_attributes__` would still work

SplitInfinity replied:
The issue is that properties are not copied to the `RecursiveScriptModule` wrapper, so they can only be used in script mode (even when they are legal; this is due to limitations with [dynamically adding properties](https://stackoverflow.com/a/1355444) but it's possible that it can be worked around). So if someone put the name of a property into `__jit_ignored_attributes__` and tried to call it, they might be confused when it doesn't work.
